crates/proto/src/errors/mod.rs

```diff
@@ -74,3 +77,65 @@ impl From<ConversionError> for tonic::Status {
         tonic::Status::invalid_argument(value.to_string())
     }
 }
+
+// GRPC ERROR TRAIT
+// ================================================================================================
+
+/// Common trait for all gRPC error enums
+pub trait GrpcError {
+    fn api_code(self) -> u8;
+    fn is_internal(&self) -> bool;
+
+    fn tonic_code(&self) -> tonic::Code {
+        if self.is_internal() {
+            tonic::Code::Internal
+        } else {
+            tonic::Code::InvalidArgument
+        }
+    }
+}
+
+// HELPER MACROS
+// ================================================================================================
+
+/// Implement the `GrpcError` trait for a given error type
+#[macro_export]
+macro_rules! grpc_error {
+    ($error:ty) => {
+        impl GrpcError for $error {
+            fn api_code(self) -> u8 {
+                self as u8
+            }
+
+            fn is_internal(&self) -> bool {
+                self == &Self::Internal
+            }
+        }
+    };
+}
+
+/// Implement Into<tonic::Status> for a given error type
+#[macro_export]
+macro_rules! into_tonic_status {
+    ($error:ty) => {
+        impl From<$error> for tonic::Status {
+            fn from(value: $error) -> Self {
+                let api_error = value.api_error();
+
+                let message = if api_error.is_internal() {
+                    "Internal error".to_owned()
+                } else {
+                    value.as_report()
+                };
+
+                tonic::Status::with_details(
+                    api_error.tonic_code(),
+                    message,
+                    vec![api_error.api_code()].into(),
+                )
+            }
+        }
+    };
+}
```

Comment:
This could be a method on `GrpcError` I think?

Reply:
Hmm not sure, `GrpcError` is implemented for this new `*GrpcError`s. While `into_tonic_status` is implemented for the higher-level errors, using the `value.as_report()` to enrich the status